mypy/checker.py

```diff
@@ -4587,6 +4763,75 @@ def or_conditional_maps(m1: TypeMap, m2: TypeMap) -> TypeMap:
     return result
 
 
+def or_partial_conditional_maps(m1: TypeMap, m2: TypeMap) -> TypeMap:
+    """Calculate what information we can learn from the truth of (e1 or e2)
+    in terms of the information that we can learn from the truth of e1 and
+    the truth of e2.
+
+    Unlike 'or_conditional_maps', we include an expression in the output even
+    if it exists in only one map: we're assuming both maps are "partial" and
+    contain information about only some expressions, and so we "or" together
+    expressions both maps have information on.
+    """
+
+    if m1 is None:
+        return m2
+    if m2 is None:
+        return m1
+    # The logic here is a blend between 'and_conditional_maps'
+    # and 'or_conditional_maps'. We use the high-level logic from the
+    # former to ensure all expressions make it in the output map,
+    # but resolve cases where both maps contain info on the same
+    # expr using the unioning strategy from the latter.
+    result = m2.copy()
+    m2_keys = {literal_hash(n2): n2 for n2 in m2}
+    for n1 in m1:
+        n2 = m2_keys.get(literal_hash(n1))
+        if n2 is None:
+            result[n1] = m1[n1]
+        else:
+            result[n2] = make_simplified_union([m1[n1], result[n2]])
+
+    return result
+
+
+def reduce_partial_type_maps(type_maps: List[Tuple[TypeMap, TypeMap]]) -> Tuple[TypeMap, TypeMap]:
+    """Reduces a list containing pairs of *partial* if/else TypeMaps into a single pair.
+
+    That is, if a expression exists in only one map, we always include it in the output.
+    We only "and"/"or" together expressions that appear in multiple if/else maps.
+
+    So for example, if we had the input:
+
+        [
+            ({x: TypeIfX, shared: TypeIfShared1}, {x: TypeElseX, shared: TypeElseShared1}),
+            ({y: TypeIfY, shared: TypeIfShared2}, {y: TypeElseY, shared: TypeElseShared2}),
+        ]
+
+    ...we'd return the output:
+
+        (
+            {x: TypeIfX,   y: TypeIfY,   shared: PseudoIntersection[TypeIfShared1, TypeIfShared2]},
+            {x: TypeElseX, y: TypeElseY, shared: Union[TypeElseShared1, TypeElseShared2]},
+        )
+
+    ...where "PseudoIntersection[X, Y] == Y" because mypy actually doesn't understand intersections
+    yet, so we settle for just arbitrarily picking the right expr's type.
```

Comment:
Do we have a test whether this would actually make a difference?

Reply:
It seems yes -- if I modify `and_conditional_maps` to bias towards picking elements from the left map, this ends up breaking the `testEnumReachabilityWithMultipleEnums` test. Basically, when we do:

```python
class Foo(Enum):
    A = 1
    B = 2
class Bar(Enum):
    A = 1
    B = 2

x3: Union[Foo, Bar]
if x3 is Foo.A or x3 is Bar.A:
    reveal_type(x3)
else:
    reveal_type(x3)
```

...we end up inferring the less precise type of `Union[Literal[Foo.B], Bar]` instead of `Union[Literal[Foo.B], Literal[Bar.B]]` in the else case -- the less precise narrowing we get from `x3 is Foo.A` overrode anything we learned in the next clause.